mlflow/store/artifact_repo.py

```diff
@@ -76,8 +89,24 @@ def download_artifacts_into(artifact_path, dest_dir):
                 self._download_file(remote_file_path=artifact_path, local_path=local_path)
             return local_path
 
-        with TempDir(remove_on_exit=False) as tmp:
-            return download_artifacts_into(artifact_path, tmp.path())
+        if dst_path is not None:
+            if not os.path.exists(dst_path):
+                raise MlflowException(
+                        message=(
+                            "The specified destination path for downloaded artifacts does not"
+                            " exist! Specified path: {dst_path}".format(dst_path=dst_path)),
+                        error_code=RESOURCE_DOES_NOT_EXIST)
+            elif not os.path.isdir(dst_path):
+                raise MlflowException(
+                        message=(
+                            "The specified destination path for downloaded artifacts must be a"
+                            " directory! Specified path: {dst_path}".format(dst_path=dst_path)),
+                        error_code=INVALID_PARAMETER_VALUE)
+
+            return download_artifacts_into(artifact_path, dst_path)
+        else:
+            with TempDir(remove_on_exit=False) as tmp:
+                return download_artifacts_into(artifact_path, tmp.path())
```

Comment:
It's a minor stylistic point, but I would expect the logic to read

"Pick a dst-path -- if one is defined, validate it, otherwise generate one. Then, download artifacts into it."

The current logic duplicates the second part, which means it doesn't define the abstraction above and increases the potential for making a mistake (e.g., if we add some logic to one branch and not the other), so it might be nice to refactor slightly.